### PR TITLE
Add support for xid in the bootstrap tool

### PR DESCRIPTION
**Pull request by pnarayanan**
Also add a gauge metric for xids in the cluster manager.

### CODE REVIEW

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/DynamicClusterManagerComponentsTest.java [hunk 1]
```diff
@@ -65,19 +66,19 @@ public DynamicClusterManagerComponentsTest() {
...
     } catch (IllegalArgumentException e) {
       // OK
     }
-    AmbryDataNode datanode1 = new AmbryDataNode("DC0", clusterMapConfig1, HOST_NAME, PORT_NUM1, RACK_ID, SSL_PORT_NUM);
-    AmbryDataNode datanode2 = new AmbryDataNode("DC1", clusterMapConfig2, HOST_NAME, PORT_NUM2, RACK_ID, SSL_PORT_NUM);
+    AmbryDataNode datanode1 = new AmbryDataNode("DC0", clusterMapConfig1, HOST_NAME, PORT_NUM1, RACK_ID, SSL_PORT_NUM, XID);
+    AmbryDataNode datanode2 = new AmbryDataNode("DC1", clusterMapConfig2, HOST_NAME, PORT_NUM2, RACK_ID, SSL_PORT_NUM, XID);
     assertEquals(datanode1.getDatacenterName(), "DC0");
     assertEquals(datanode1.getHostname(), HOST_NAME);
     assertEquals(datanode1.getPort(), PORT_NUM1);
```
Comment (jsjtzyy) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/DynamicClusterManagerComponentsTest.java]: missing the test `assertEquals(datanode1.getXid(), XID);`

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java [hunk 2]
```diff
@@ -200,7 +202,8 @@ static JSONArray getJsonArrayDataNodesRackAware(int dataNodeCount, String hostna
   private static void updateJsonArrayDataNodesRackAware(JSONArray dataNodeJsonArray, int dataNodeCount, String hostname,
       int basePort, int sslPort, int numRacks, HardwareState hardwareState, JSONArray disks) throws JSONException {
     for (int i = dataNodeJsonArray.length(); i < dataNodeCount; ++i) {
-      dataNodeJsonArray.put(getJsonDataNode(hostname, basePort + i, sslPort + i, i % numRacks, hardwareState, disks));
+      dataNodeJsonArray.put(
+          getJsonDataNode(hostname, basePort + i, sslPort + i, i % numRacks, 64, hardwareState, disks));
```
Comment (jsjtzyy) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java]: why not use a constant value at the beginning of this file? Then we can apply it in different places.

**Comment by jsjtzyy** [comment]
I didn't see any relevant test in `HelixBootstrapUpgradeTooTest`, is it feasible to have a test?

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/DynamicClusterManagerComponentsTest.java (hunk 1 above)
Comment (pnarayanan) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/DynamicClusterManagerComponentsTest.java]: Added.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java [hunk 3]
```diff
@@ -51,13 +52,14 @@
    * @throws Exception if there is an exception in instantiating the {@link ResourceStatePolicy}
    */
   AmbryDataNode(String dataCenterName, ClusterMapConfig clusterMapConfig, String hostName, int portNum, String rackId,
-      Integer sslPortNum) throws Exception {
+      Integer sslPortNum, String xid) throws Exception {
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryDatanode.java]: Done.

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java (hunk 2 above)
Comment (pnarayanan) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java]: Done.

**Comment by pnarayanan** [comment]
HelixBootstrapTool does a validation at the end, and now it also validates if the xid in the json files is the same as the xid obtained when the instance configs are queried.

The tool's test is limited to this functionality since it does not concern itself with the logic around how xids are handled. For the logic we have tests elsewhere (which we added in the previous PR).

I will make a small change though in TestUtils so that it sets a valid non-null xid when datanodes are created, so that the tool's validation exercises the non-null case.

**Comment by pnarayanan** [comment]
@jsjtzyy addressed your comments.

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/DataNodeTest.java [hunk 4]
```diff
@@ -101,16 +101,18 @@ public void basics() throws JSONException {
...
     assertEquals(dataNode.toJSONObject().toString(), jsonObject.toString());
     assertEquals(dataNode, new TestDataNode("datacenter", dataNode.toJSONObject(), clusterMapConfig));
 
     // Test with defined rackId
     jsonObject =
-        TestUtils.getJsonDataNode(TestUtils.getLocalHost(), 6666, 7666, 42, HardwareState.AVAILABLE, getDisks());
+        TestUtils.getJsonDataNode(TestUtils.getLocalHost(), 6666, 7666, 42, 64, HardwareState.AVAILABLE, getDisks());
     dataNode = new TestDataNode("datacenter", jsonObject, clusterMapConfig);
     assertEquals("42", dataNode.getRackId());
+    assertEquals("64", dataNode.getXid());
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/DataNodeTest.java]: Done.

**Comment by pnarayanan** [comment]
@jsjtzyy Addressed comments, also made the change we discussed yesterday to ignore xids for changes to self; added a test.

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java [hunk 5]
```diff
@@ -514,6 +514,13 @@ public void xidTest() throws Exception {
       assertFalse(instanceName.equals(aheadInstanceConfig.getInstanceName()));
     }
 
+    // Ahead instance should be honored if the cluster manager is of the aheadInstance.
+    HelixClusterManager aheadInstanceClusterManager =
+        new HelixClusterManager(clusterMapConfig, aheadInstanceConfig.getInstanceName(), helixManagerFactory,
+            new MetricRegistry());
+    assertEquals(instanceCount, aheadInstanceClusterManager.getDataNodeIds().size());
+    aheadInstanceClusterManager.close();
+
```
Comment (jsjtzyy) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java]: Make sure the `clusterManager` is closed at the end of this test.

Comment (pnarayanan) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java]: Ah, good catch. The intention was to use the member variable named `clusterManager` (which gets closed automatically in @After). Made changes to do that.

**Comment by pnarayanan** [comment]
@jsjtzyy @cgtz made the change to `getXid()` to return long instead of String; rebased and fixed merged conflicts.

**Pull request merged**